### PR TITLE
MGMT-4261 Agent CR and ACI cleanup when deleting CD

### PR DESCRIPTION
**Pull request by nmagnezi**
clusterdeployments_controller:
When a CD CR gets deleted, the controller will place a
finalizer to trigger a pre-deletion cleanup.

The cleanup includes ACI deletion, which will place a finalizer on ACI.

ACI finalizer includes backend-cluster deregister and deletion of the
cluster agent resources.

As for subsystem tests:
1. Tests will now assert for agent deletion per ACI removal.
2. Tests will now assert for agent deletion and ACI per CD removal.
3. Each test will now use a unique CD and ACI CR name - to provide better
   test isolation. This was also done because deletion via finalizers
   takes an additional reconcile to be fully deleted.
4. Per 2, and to avoid any cleanup bugs, the cleanup done for subsystem
   tests (AfterEach) is now followed up by verifyCleanUP

P.S.
Note that this PR does not cover host deregister upon Agent CR deletion,
which was added in #1642

### CODE REVIEW

**Comment by nmagnezi** [comment]
/hold
adding tests

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (mhrivnak) [inline internal/controller/controllers/clusterdeployments_controller.go:141]: I'm not understanding how this solves the problem of identifying when a ClusterDeployment has been deleted. Can you describe the pattern?

**Review by mhrivnak — CHANGES_REQUESTED** [review]

File: internal/controller/controllers/agent_controller.go
Comment (dgoodwin) [reply internal/controller/controllers/agent_controller.go:133]: controller runtime does track metrics for error rates, so in theory if you'd like this problem to show up in those (as you might have alerts on them), it might be good to return an error if you're requeuing because of an error.

File: internal/controller/controllers/agent_controller.go
Comment (dgoodwin) [inline internal/controller/controllers/agent_controller.go:86]: Typically if you have cleanup to do the best practice is to make sure there's a finalizer on clusterdeployment in this case (that will change shortly, see below), and when you see ClusterDeployment with deletionTimestamp set and that finalizer still present, you know you need to take action.

This might be fine but there are some weird cases I can think of. Hive supports cluster migration from one Hive shard to another, in which case if the ClusterDeployment were gone, Agents would start deleting.
Or, it may also proc if you are creating Agents with a ref to a ClusterDeployment that doesn't exist yet, (maybe it's further down in file of yaml docs or a kubectl apply command). 

Just wanted to suggest the finalizer approach as that should be a more reliable way to handle this. This might be fine as is in which case disregard.

Once the AgentClusterInstall PR merges the place for such a finalizer would be on that resource, not ClusterDeployment.

Comment (mhrivnak) [reply internal/controller/controllers/agent_controller.go:86]: Totally agree. I was just waiting for Nir to elaborate on how this code is expected to work as-is before bringing up finalizers, but I agree this almost certainly calls for using one.

File: internal/controller/controllers/agent_controller.go
Comment (mhrivnak) [reply internal/controller/controllers/agent_controller.go:133]: Good point. I tend to think that metric isn't particularly useful, but it's subjective. Here's a longer discussion on the topic: https://github.com/kubernetes-sigs/controller-runtime/issues/617

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (nmagnezi) [reply internal/controller/controllers/clusterdeployments_controller.go:141]: Allow me to explain:
The basic idea I had in mind is that we should trigger an agent reconcile in some cases (mentioned in the commit-msg) so the `agent_controller` can determine whether or not the resource should be deleted.
Since those cases start with the `clusterDeployment`, it will naturally first reconcile on that controller, so it needs to "tell" the `agent_controller` that something happened.

The outcome is that I placed (in `clusterdeployments_controller`) a call to notify `agent_controller` about relevant agents in places that indicate the day1 cluster got deregistered (line#135) and on `deregisterClusterIfNeeded()` (line#587), which gets invoked when it gets `k8serrors.IsNotFound`
while trying to get the `clusterDeployment` CR.

On the agent controller side, it starts to reconcile and call `r.deleteAgent` if either `clusterDeployment` or backend cluster is gone; or that agent does not belong to the cluster it is listed for (day2 scenario).

Let me know if that clarifies the current implementation, and how to proceed from here.

File: internal/controller/controllers/agent_controller.go
Comment (nmagnezi) [reply internal/controller/controllers/agent_controller.go:86]: Can you please point me to a good reference re: finalizers? in case we decide to go with that (just replied on the other comment).

File: internal/controller/controllers/agent_controller.go
Comment (nmagnezi) [reply internal/controller/controllers/agent_controller.go]: Changed to return `nil` and Requeue set `true` only if we had an error deleting that agent.

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (mhrivnak) [reply internal/controller/controllers/clusterdeployments_controller.go:141]: Thanks for elaborating. The k8s-native way to normally handle this situation is to add a finalizer to the ClusterDeployment. Then at the beginning of the Reconcile, check for a deletion timestamp to be set, and if it is and the finalizer is still there:
* send the reconcile requests to the agent controller
* do any other things you want to do upon deletion
* remove the finalizer and save the resource
* return

I know in our case we have this extra source of truth in the sql DB, but I think it's better to not depend on that state machine as a substitute for the normal kube patterns.

WDYT?

Comment (filanov) [reply internal/controller/controllers/clusterdeployments_controller.go:141]: Regardless of finilazer  - i don't think that you need to notify the agents when cluster is deleted there is a deletion of the agents in the backend so there should be an event for each agent and it will trigger the reconcile for the agents.

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (filanov) [inline internal/controller/controllers/clusterdeployments_controller.go:587]: same as above IMO notification should come from the backend.

File: internal/controller/controllers/agent_controller.go
Comment (filanov) [reply internal/controller/controllers/agent_controller.go:86]: https://book.kubebuilder.io/reference/using-finalizers.html

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (filanov) [inline internal/controller/controllers/clusterdeployments_controller.go:60]: Doesn't it defined in Hive? can't you import this definition?

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (filanov) [inline internal/controller/controllers/clusterdeployments_controller.go:101]: Did you test it with Hive? trying to understand if there will be a collision with Hive

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (nmagnezi) [reply internal/controller/controllers/clusterdeployments_controller.go]: Implementation changed to use a finalizer for clusterDeployment deletion.

the cleanup logic `deregisterClusterIfNeeded` is being called to cleanup the backend cluster and delete the agents.

That very same cleanup logic is used when (not changed in this PR):
1. It tries to reconcile for a non-existent `clusterDeployment` - to ensure both the controller and the backend are on the same page.
2. Day1 Cluster gets deleted, which we know won't lead to a `clusterDeployment` removal but to creating a day2 cluster.

File: internal/controller/controllers/agent_controller.go
Comment (nmagnezi) [reply internal/controller/controllers/agent_controller.go]: resolving this comment, since this is no longer a part of this PR.
to be included in a follow-up PR that handles host deregister.

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (filanov) [inline internal/controller/controllers/clusterdeployments_controller.go:910]: you can use `funk.Contains` no need to define this function.

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (nmagnezi) [reply internal/controller/controllers/clusterdeployments_controller.go:60]: Looked it up and I could not find it in Hive.
@dgoodwin, did I miss it?

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (nmagnezi) [reply internal/controller/controllers/clusterdeployments_controller.go]: Done

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (nmagnezi) [reply internal/controller/controllers/clusterdeployments_controller.go]: Per discussions, we had on slack: will not use backend notification for this.

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (nmagnezi) [reply internal/controller/controllers/clusterdeployments_controller.go:102]: Good point. I'll look into it.

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (dgoodwin) [reply internal/controller/controllers/clusterdeployments_controller.go:61]: You'll want to use your own finalizer, as we would never want two controllers that might be trying to remove the same finalizer. Bear in mind that this finalizer is about to move to the AgentClusterInstall once Fred's work merges.

I'd do something like agent-install.openshift.io/deprovision

**Comment by nmagnezi** [comment]
/hold
looking into `ci/prow/subsystem-kubeapi-aws` results.
Feel free to provide feedback, regardless.

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (mhrivnak) [inline internal/controller/controllers/clusterdeployments_controller.go:125]: Typically a Reconcile function would return after doing this. Further actions, like trying to update status, would fail. If there's anything that happens below that's relevant to cleanup, it should probably happen in this block prior to removing the finalizer.

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (mhrivnak) [inline internal/controller/controllers/clusterdeployments_controller.go:717]: You can use the `ctx` you passed into this function.

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (mhrivnak) [inline internal/controller/controllers/clusterdeployments_controller.go:723]: The agent namespace is also useful. Since this is a lot of info, maybe put the identifiers for the resources into fields on the logger. Or you could at least do that at the top of this function for the CD, and then identify the Agent in the message if you want.

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (mhrivnak) [inline internal/controller/controllers/clusterdeployments_controller.go:725]: Can you just use `clusterAgent` instead of instantiating a new Agent?

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (mhrivnak) [inline internal/controller/controllers/clusterdeployments_controller.go:86]: Can you explain this change? `ClusterDeployment` doesn't exist in API group `agent-install.openshift.io`. And what's up with the `deprovision` subresource?

I'm thinking this line was correct before this change.

**Review by mhrivnak — CHANGES_REQUESTED** [review]

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (nmagnezi) [reply internal/controller/controllers/clusterdeployments_controller.go]: Done

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (nmagnezi) [reply internal/controller/controllers/clusterdeployments_controller.go]: Done

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (nmagnezi) [reply internal/controller/controllers/clusterdeployments_controller.go]: I changed this as a follow-up to the proposal @dgoodwin made in https://github.com/openshift/assisted-service/pull/1606#discussion_r625018140

You are correct to mention that the API group should remain `hive.openshift.io`, so I reverted back to that.
As far as the finalizer name goes, I see `deprovision` [is mentioned in Hive code base](https://github.com/openshift/hive/blob/5c038fb051906a7c2d733e7b017690044c66c124/apis/hive/v1/clusterdeployment_types.go#L21-L23) and `finalizers` seems like the generic name (at least [according to the docs](https://book.kubebuilder.io/reference/using-finalizers.html)).
I proposed `ai-deprovision`, to make it a finalized specific to `assisted-installer`.
let me know what you think about it.

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (nmagnezi) [reply internal/controller/controllers/clusterdeployments_controller.go]: Done

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (nmagnezi) [reply internal/controller/controllers/clusterdeployments_controller.go:125]: Done

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (mhrivnak) [reply internal/controller/controllers/clusterdeployments_controller.go]: This is not the place to put the name of a specific finalizer. It refers to a user's ability to modify the `finalizers` field on a resource's `ObjectMeta`.

Comment (mhrivnak) [reply internal/controller/controllers/clusterdeployments_controller.go]: You can see the example here: https://book.kubebuilder.io/reference/using-finalizers.html

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (filanov) [inline internal/controller/controllers/clusterdeployments_controller.go:109]: Once we add an owner reference we can probably look at only agent cluster install no?

Comment (nmagnezi) [reply internal/controller/controllers/clusterdeployments_controller.go:109]: I guess the question we should ask: when do we want to have that cleanup triggered? on `clusterDeployment` deletion or `agentClusterInstall` deletion (which may happen either with that resource deletion or when we delete `clusterDeployment` because of the ownerRef)

Comment (filanov) [reply internal/controller/controllers/clusterdeployments_controller.go:109]: Actually infra env reference to cluster deployment and not to agent cluster install so we don't really have a way to enforce it.

Comment (nmagnezi) [reply internal/controller/controllers/clusterdeployments_controller.go:109]: I took a second look at the code that was added with `AgentClusterInstall`.
We stop the reconcile in case there is no `AgentClusterInstall` to be found for the `clusterDeployment`, meaning we will not call `RegisterClusterInternal()`

https://github.com/openshift/assisted-service/blob/d3657fd59d41d5a9301bf96d2adea27974763a8d/internal/controller/controllers/clusterdeployments_controller.go#L106-L110

So maybe it does make sense to deregister (and delete agents) when `AgentClusterInstall` gets removed? seems like a more "symmetrical". what do you think?

BTW, not sure how that'll play out with the day2 flow, I need to look at that as well.

Comment (dgoodwin) [reply internal/controller/controllers/clusterdeployments_controller.go:109]: I would trigger cleanup on AgentClusterInstall rather than ClusterDeployment.

Comment (nmagnezi) [reply internal/controller/controllers/clusterdeployments_controller.go:109]: @dgoodwin for what reason?

Comment (dgoodwin) [reply internal/controller/controllers/clusterdeployments_controller.go:109]: Teardown logic is intricately linked to the provision logic, so it makes the most sense for the finalizers to be tied to implementation specific resource. I am also hoping cluster install controllers *never* write to the ClusterDeployment, and hive controllers never write to an external cluster install CR.

Comment (filanov) [reply internal/controller/controllers/clusterdeployments_controller.go:109]: We reconcile on cluster deployment as the main resource, so once it was deleted and we need to do the cleanup the reconcile loop will not find cluster deployment so how should it work?

Comment (dgoodwin) [reply internal/controller/controllers/clusterdeployments_controller.go:109]: Hrm, if your AgentClusterInstall has a finalizer, and an OwnerRef to ClusterDeployment, will the ClusterDeployment disappear before your AgentClusterInstall? I am thinking it will remain.

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (filanov) [inline internal/controller/controllers/clusterdeployments_controller.go:99]: maybe log the error it will be clearer.

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (filanov) [inline internal/controller/controllers/clusterdeployments_controller.go:136]: when can it be true? in line 24 you set this var to be true but then you just return, so maybe this parameters is not needed?

Comment (nmagnezi) [reply internal/controller/controllers/clusterdeployments_controller.go:136]: `clusterInstallDeleted` is set  to `true` when there is no ACI to fetch (line 122), but will return only if `clusterDeployment` is not being deleted (line 128).

The other return statement in that block is in a case that we failed to fetch ACI for other reasons (line 132)

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (nmagnezi) [reply internal/controller/controllers/clusterdeployments_controller.go:99]: ack

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (filanov) [reply internal/controller/controllers/clusterdeployments_controller.go:136]: missed that part thanks.

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (nmagnezi) [reply internal/controller/controllers/clusterdeployments_controller.go:99]: done

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (nmagnezi) [reply internal/controller/controllers/clusterdeployments_controller.go:64]: I used our group name to indicate that, but maybe we can also rename the finalized to `ai-deprovision`, same as in 

https://github.com/openshift/assisted-service/blob/38f11f4d741acfff11db1d76a757b836f63bd4a8/internal/controller/controllers/agent_controller.go#L53

Comment (filanov) [reply internal/controller/controllers/clusterdeployments_controller.go:64]: Yes, I think that it will be clearer.

Comment (nmagnezi) [reply internal/controller/controllers/clusterdeployments_controller.go]: done

**Review by filanov — APPROVED** [review]

**Comment by nmagnezi** [comment]
> New changes are detected. LGTM label has been removed.

Rebased to resolve conflicts.

**Review by filanov — APPROVED** [review]

**Pull request merged**
